x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/CompositeRolesStore.java

```diff
@@ -152,88 +171,113 @@ public void roles(Set<String> roleNames, FieldPermissionsCache fieldPermissionsC
         }
     }
 
-    private void roleDescriptors(Set<String> roleNames, ActionListener<Set<RoleDescriptor>> roleDescriptorActionListener) {
+    private void roleDescriptors(Set<String> roleNames, ActionListener<RolesRetrievalResult> rolesResultListener) {
         final Set<String> filteredRoleNames = roleNames.stream().filter((s) -> {
-            if (negativeLookupCache.contains(s)) {
+            if (negativeLookupCache.get(s) != null) {
                 logger.debug("Requested role [{}] does not exist (cached)", s);
                 return false;
             } else {
                 return true;
             }
         }).collect(Collectors.toSet());
+        final RolesRetrievalResult retrievalResult = new RolesRetrievalResult();
         final Set<RoleDescriptor> builtInRoleDescriptors = getBuiltInRoleDescriptors(filteredRoleNames);
+        retrievalResult.addDescriptors(builtInRoleDescriptors);
         Set<String> remainingRoleNames = difference(filteredRoleNames, builtInRoleDescriptors);
+
         if (remainingRoleNames.isEmpty()) {
-            roleDescriptorActionListener.onResponse(Collections.unmodifiableSet(builtInRoleDescriptors));
+            rolesResultListener.onResponse(retrievalResult);
         } else {
-            nativeRolesStore.getRoleDescriptors(remainingRoleNames.toArray(Strings.EMPTY_ARRAY), ActionListener.wrap((descriptors) -> {
-                logger.debug(() -> new ParameterizedMessage("Roles [{}] were resolved from the native index store", names(descriptors)));
-                builtInRoleDescriptors.addAll(descriptors);
-                callCustomRoleProvidersIfEnabled(builtInRoleDescriptors, filteredRoleNames, roleDescriptorActionListener);
-            }, e -> {
-                logger.warn("role retrieval failed from the native roles store", e);
-                callCustomRoleProvidersIfEnabled(builtInRoleDescriptors, filteredRoleNames, roleDescriptorActionListener);
-            }));
+            final Set<RoleDescriptor> fileRoleDescriptors = getDescriptorsFromFileStore(remainingRoleNames);
+            retrievalResult.addDescriptors(fileRoleDescriptors);
```

Comment:
I didn't add a interface for this but instead just used `BiConsumer<Set<String>, ActionListener<RoleRetrievalResult>>` since we have that for custom role providers.